stan/math/rev/functor/cvodes_integrator_adjoint.hpp

```diff
@@ -0,0 +1,829 @@
+        CVodeSVtolerances(solver_->cvodes_mem_, relative_tolerance_forward_,
+                          solver_->nv_absolute_tolerance_forward_),
+        "CVodeSVtolerances");
+
+    check_flag_sundials(
+        CVodeSetLinearSolver(solver_->cvodes_mem_, solver_->LS_forward_,
+                             solver_->A_forward_),
+        "CVodeSetLinearSolver");
+
+    check_flag_sundials(
+        CVodeSetJacFn(solver_->cvodes_mem_,
+                      &cvodes_integrator_adjoint_vari::cv_jacobian_states),
+        "CVodeSetJacFn");
+
+    // initialize backward sensitivity system of CVODES as needed
+    if (is_var_return) {
+      check_flag_sundials(
+          CVodeAdjInit(solver_->cvodes_mem_, num_steps_between_checkpoints_,
+                       interpolation_polynomial_),
+          "CVodeAdjInit");
+    }
+  }
+
+  /**
+   * Solve the ODE initial value problem y' = f(t, y), y(t0) = y0 at a set of
+   * times, { t1, t2, t3, ... } using the stiff backward differentiation formula
+   * (BDF) solver in CVODES.
+   *
+   * @return std::vector of Eigen::Matrix of the states of the ODE, one for each
+   *   solution time (excluding the initial state)
+   */
+  std::vector<Eigen::Matrix<T_Return, Eigen::Dynamic, 1>> operator()() {
+    const double t0_dbl = value_of(t0_);
+    const auto ts_dbl = value_of(ts_);
+    std::vector<Eigen::Matrix<T_Return, Eigen::Dynamic, 1>> y_return(
+        ts_.size());
+
+    double t_init = t0_dbl;
+    for (size_t n = 0; n < ts_dbl.size(); ++n) {
+      double t_final = ts_dbl[n];
+      if (t_final != t_init) {
+        if (is_var_return) {
+          int ncheck;
+
+          int error_code
+              = CVodeF(solver_->cvodes_mem_, t_final,
+                       solver_->nv_state_forward_, &t_init, CV_NORMAL, &ncheck);
+
+          if (error_code == CV_TOO_MUCH_WORK) {
+            throw_domain_error(solver_->function_name_, "", t_final,
+                               "Failed to integrate to next output time (",
+                               ") in less than max_num_steps steps");
+          } else {
+            check_flag_sundials(error_code, "CVodeF");
+          }
+
+        } else {
+          int error_code
+              = CVode(solver_->cvodes_mem_, t_final, solver_->nv_state_forward_,
+                      &t_init, CV_NORMAL);
+
+          if (error_code == CV_TOO_MUCH_WORK) {
+            throw_domain_error(solver_->function_name_, "", t_final,
+                               "Failed to integrate to next output time (",
+                               ") in less than max_num_steps steps");
+          } else {
+            check_flag_sundials(error_code, "CVode");
+          }
+        }
+      }
+      store_state(n, state_forward_, y_return[n]);
+
+      t_init = t_final;
+    }
+
+    forward_returned_ = true;
+    return y_return;
+  }
+
+  virtual void chain() {
+    if (solver_->cvodes_mem_ == nullptr) {
+      return;
+    }
+
+    if (!forward_returned_) {
+      return;
+    }
```

Comment:
Is there ever a case where `operator()()` won't be called / `forward_returned_` won't be true by the time we call `chain()`?

Reply:
yeah... these if's can be reviewed. Also: How about we move the code inside the `operator()()` into the constructor? The forward solution must be calculated to make any sense for this vari.